File: benchmarking/scenarios/inference-rest.js
import { readTestData } from "../common/helpers.js";
import { RestClient } from "../common/rest.js";
import { Trend, Counter, Gauge } from "k6/metrics";

const TestData = {
  iris: readTestData("iris"),
};

const rest = new RestClient();

const ScenarioDuration = "60s";
const ScenarioVUs = 300;

export const options = {
  scenarios: {
    iris_rest: {
      executor: "constant-vus",
      duration: ScenarioDuration,
      vus: ScenarioVUs,
      tags: { model_name: "iris", protocol: "rest" },
      env: { MODEL_NAME: "iris", PROTOCOL: "rest" },
    },
  },
  thresholds: {
    http_reqs: ["rate > 1100"],
  },
};

// Worker-aware unary metrics
const unaryWorkerReqs = new Counter("unary_worker_reqs");
const unaryWorkerProcRssKb = new Trend("unary_worker_proc_rss_kb");
const unaryWorkerProcCpuPct = new Trend("unary_worker_proc_cpu_pct");
const unaryWorkersSeenLocal = new Trend("unary_workers_seen_local");
const unaryDbgWorkersSeen = new Gauge("unary_debug_workers_seen_last_batch");

export function setup() {
  rest.loadModel("iris");
  return TestData;
}

export default function (data) {
  const modelName = __ENV.MODEL_NAME;

  const res = rest.infer(modelName, data[modelName].rest);

  const seenWorkers = new Set();

  try {
    const body = JSON.parse(res.body);
    const outputs = body.outputs || [];
    const params = outputs[0]?.parameters || {};
    const headers = params.headers || {};
    const pid = headers["X-Worker-PID"] || "unknown";
    seenWorkers.add(pid);
    unaryWorkerReqs.add(1, { worker: pid, model_name: modelName });

    const rss = Number(headers["X-Proc-RSS-KB"]);
    if (!Number.isNaN(rss)) {
      unaryWorkerProcRssKb.add(rss, { worker: pid, model_name: modelName });
    }
    const cpu = Number(headers["X-Proc-CPU-Pct"]);
    if (!Number.isNaN(cpu)) {
      unaryWorkerProcCpuPct.add(cpu, { worker: pid, model_name: modelName });
    }
  } catch (_) {
    // ignore parse issues
  }

  unaryWorkersSeenLocal.add(seenWorkers.size, { model_name: modelName });
  unaryDbgWorkersSeen.add(seenWorkers.size, { model_name: modelName });
}

export function teardown(data) {
  rest.unloadModel("iris");
}

/* ------------------------- Summary (REST unary) ------------------------- */

export function handleSummary(data) {

  const lines = [];
  lines.push("=== REST Unary Summary ===");
  lines.push(`Approx. distinct workers (max per-VU batch): ${approxWorkers}`);
  if (avgRss !== null) lines.push(`Average worker RSS (KB): ${avgRss.toFixed(0)}`);
  if (avgCpu !== null) lines.push(`Average worker CPU (%): ${avgCpu.toFixed(2)}`);
  lines.push("");

  // Classic k6 table
  const classic = textSummary(data, { indent: " ", enableColors: true });

  // Single stdout output containing both sections
  return { stdout: `${lines.join("\n")}\n${classic}` };
}

File: benchmarking/scenarios/inference_rest.py
import json
import os

from common.helpers import read_test_data
from locust import HttpUser, between, events, task

# Config (mirror k6)
SCENARIO_DURATION = "60s"  # for docs
SCENARIO_VUS = 300         # run via CLI: -u 300 -t 60s

TestData = {
    "iris": read_test_data("iris"),
}

def _parse_worker_headers_from_json_body(text):
    try:
        body = json.loads(text)
        outputs = body.get("outputs", [])
        params = (outputs[0] or {}).get("parameters", {}) if outputs else {}
        headers = params.get("headers", {}) or {}
        return headers
    except Exception:
        return {}

class RestIrisUser(HttpUser):
    wait_time = between(0.0, 0.0)
    host = f"http://{os.environ.get('MLSERVER_HOST','127.0.0.1')}:{os.environ.get('MLSERVER_HTTP_PORT','8080')}"

    def on_start(self):
        # Best-effort model load
        self.client.post("/v2/repository/models/iris/load")

    def on_stop(self):
        self.client.post("/v2/repository/models/iris/unload")

    @task
    def iris_infer(self):
        payload = TestData["iris"]["rest"]
        with self.client.post("/v2/models/iris/infer", json=payload, name="REST /infer iris", catch_response=True) as r:
            if r.status_code != 200:
                r.failure(f"HTTP {r.status_code}")
                return
            headers = _parse_worker_headers_from_json_body(r.text)
            pid = headers.get("X-Worker-PID", "unknown")
            # Emit custom "per-worker" metrics by piggybacking on request events:
            if "X-Proc-RSS-KB" in headers:
                try:
                    rss = float(headers["X-Proc-RSS-KB"])
                    events.request.fire(request_type="metric", name="unary_worker_proc_rss_kb", response_time=rss, response_length=0, context={"worker": pid}, exception=None)
                except Exception:
                    pass
            if "X-Proc-CPU-Pct" in headers:
                try:
                    cpu = float(headers["X-Proc-CPU-Pct"])
                    events.request.fire(request_type="metric", name="unary_worker_proc_cpu_pct", response_time=cpu, response_length=0, context={"worker": pid}, exception=None)
                except Exception:
                    pass
            # Distinct workers seen per-request (will usually be 1 for unary)
            events.request.fire(request_type="metric", name="unary_workers_seen_local", response_time=1.0, response_length=0, context={"workers": 1}, exception=None)
